// deep-scope-analyser/lib/scope/moduleScope.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const scope_1 = require("./scope");
const importManager_1 = require("../importManager");
const exportManager_1 = require("../exportManager");
class ModuleScope extends scope_1.Scope {
    constructor(scopeManager, upperScope, block) {
        super(scopeManager, "module", upperScope, block, false);
        this.importManager = new importManager_1.ImportManager();
        this.exportManager = new exportManager_1.ExportManager();
        this.isExportingNamedDeclaration = false;
    }
    __define(node, def) {
        const ancestor = super.__define(node, def);
        if (ancestor !== null && this.isExportingNamedDeclaration) {
            this.exportManager.addLocalExportVariable({
                type: exportManager_1.ExportVariableType.Local,
                exportName: ancestor.name,
                localName: ancestor.name,
                node: ancestor.defs[0].node,
            });
        }
        return ancestor;
    }
}
exports.ModuleScope = ModuleScope;
