// deep-scope-analyser/lib/exportManager.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var ExternalType;
(function (ExternalType) {
    ExternalType["Identifier"] = "identifier";
    ExternalType["All"] = "all";
})(ExternalType = exports.ExternalType || (exports.ExternalType = {}));
var ExportVariableType;
(function (ExportVariableType) {
    ExportVariableType["Local"] = "local";
    ExportVariableType["External"] = "external";
})(ExportVariableType = exports.ExportVariableType || (exports.ExportVariableType = {}));
class ExportManager {
    constructor() {
        this.exportsMap = new Map();
        this.localVariables = [];
        this.externalVariables = [];
        this.exportDefaultDeclaration = null;
    }
    addLocalExportVariable(exportVar) {
        this.exportsMap.set(exportVar.exportName, exportVar);
        this.localVariables.push(exportVar);
        if (exportVar.exportName === "default") {
            this.exportDefaultDeclaration = exportVar.node;
        }
    }
    addExternalVariable(external) {
        this.externalVariables.push(external);
        if (external.moduleType === ExternalType.Identifier) {
            this.exportsMap.set(external.names.exportName, external);
        }
    }
}
exports.ExportManager = ExportManager;
