// deep-scope-analyser/lib/reference.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const __READ = 0x1;
const __WRITE = 0x2;
const __RW = __READ | __WRITE;
const __EXPORT = 0x4;
class Reference {
    constructor(identifier, from, flag, writeExpr, maybeImplicitGlobal = null, partial, init) {
        this.identifier = identifier;
        this.from = from;
        this.flag = flag;
        this.maybeImplicitGlobal = maybeImplicitGlobal;
        this.tainted = false;
        this.tainted = false;
        this.resolved = null;
        this.flag = flag;
        if (this.isWrite) {
            this.writeExpr = writeExpr;
            this.partial = partial;
            this.init = init;
        }
    }
    get isStatic() {
        return (!this.tainted &&
            this.resolved &&
            this.resolved.scope.isStatic());
    }
    get isWrite() {
        return !!(this.flag & Reference.WRITE);
    }
    get isRead() {
        return !!(this.flag & Reference.READ);
    }
    get isReadOnly() {
        return this.flag === Reference.READ;
    }
    get isWriteOnly() {
        return this.flag === Reference.WRITE;
    }
    get isReadWrite() {
        return this.flag === Reference.RW;
    }
    get isExport() {
        return this.flag === Reference.EXPORT;
    }
}
Reference.READ = __READ;
Reference.WRITE = __WRITE;
Reference.RW = __RW;
Reference.EXPORT = __EXPORT;
exports.Reference = Reference;
