// deep-scope-analyser/lib/analyser/childScopesTraverser.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const estraverse = require("estraverse");
class PureDeclaratorTraverser {
    constructor(validatorDeclarator, moduleScope) {
        this.validatorDeclarator = validatorDeclarator;
        this.moduleScope = moduleScope;
        this.refsToModule = [];
        this.relevantScopes = [];
        this.ids = [];
        moduleScope.childScopes.forEach(scope => {
            const block = scope.block;
            if (this.nodeContains(validatorDeclarator, block)) {
                this.relevantScopes.push(scope);
            }
        });
        estraverse.traverse(validatorDeclarator, {
            enter: (node) => {
                if (node.type === "Identifier") {
                    this.ids.push(node);
                    const idName = node.name;
                    this.refsToModule.push(idName);
                }
            },
        });
    }
    nodeContains(node1, node2) {
        return node2.start >= node1.start && node2.end <= node1.end;
    }
}
exports.PureDeclaratorTraverser = PureDeclaratorTraverser;
class ChildScopesTraverser {
    constructor(scope) {
        this.scope = scope;
        this.refsToModule = [];
        this.traverse = (scope) => {
            scope.references.forEach(ref => {
                if (ref.resolved &&
                    ref.resolved.scope.type === "module") {
                    this.refsToModule.push(ref.identifier.name);
                }
            });
            scope.childScopes.forEach(this.traverse);
        };
        this.traverse(scope);
    }
}
exports.ChildScopesTraverser = ChildScopesTraverser;
