// deep-scope-analyser/lib/analyser/virtualScope/index.js
"use strict";
function __export(m) {
    for (var p in m) if (!exports.hasOwnProperty(p)) exports[p] = m[p];
}
Object.defineProperty(exports, "__esModule", { value: true });
var VScopeContentType;
(function (VScopeContentType) {
    VScopeContentType["Import"] = "import";
    VScopeContentType["Reference"] = "reference";
    VScopeContentType["ArrowFunction"] = "arrow-function";
    VScopeContentType["ClassExpression"] = "class-expression";
    VScopeContentType["ClassDeclaration"] = "class-declaration";
    VScopeContentType["PureFunctionCall"] = "pure-function-call";
    VScopeContentType["NormalFunctionCall"] = "normal-function-call";
    VScopeContentType["FunctionExpression"] = "function-expression";
    VScopeContentType["FunctionDeclaration"] = "function-declaration";
    VScopeContentType["Undefined"] = "undefined";
})(VScopeContentType = exports.VScopeContentType || (exports.VScopeContentType = {}));
var VirtualScopeType;
(function (VirtualScopeType) {
    VirtualScopeType["Variable"] = "variable";
    VirtualScopeType["Default"] = "default";
})(VirtualScopeType = exports.VirtualScopeType || (exports.VirtualScopeType = {}));
__export(require("./variableVirtualScope"));
__export(require("./exportDefaultVirtualScope"));
